code/game/objects/structures/tank_dispenser.dm

```diff
@@ -1,117 +1,135 @@
-				if(oxygentanks < 5)
-					update_icon()
-			else
-				to_chat(user, "<span class='notice'>[src] is full.</span>")
-			updateUsrDialog()
-			return
-		if(istype(I, /obj/item/tank/hydrogen))
-			if(hydrogentanks < 10)
-				if(!user.unEquip(I, src))
-					return
-				hydtanks.Add(I)
-				hydrogentanks++
-				to_chat(user, "<span class='notice'>You put [I] in [src].</span>")
-				if(oxygentanks < 6)
-					update_icon()
-			else
-				to_chat(user, "<span class='notice'>[src] is full.</span>")
-			updateUsrDialog()
-			return
-
-/obj/structure/dispenser/Topic(href, href_list)
-	if(usr.stat || usr.restrained())
-		return
-	if(Adjacent(usr))
-		usr.set_machine(src)
-		if(href_list["oxygen"])
-			if(oxygentanks > 0)
-				var/obj/item/tank/oxygen/O
-				if(oxytanks.len == oxygentanks)
-					O = oxytanks[1]
-					oxytanks.Remove(O)
-				else
-					O = new /obj/item/tank/oxygen(loc)
-				O.dropInto(loc)
-				to_chat(usr, "<span class='notice'>You take [O] out of [src].</span>")
-				oxygentanks--
+	if(istype(I, /obj/item/tank/oxygen) || istype(I, /obj/item/tank/air))
+		if(LAZYLEN(oxygen_tanks) < 10)
+			if(!user.unEquip(I, src))
+				return
+			LAZYADD(oxygen_tanks, I)
+			to_chat(user, SPAN_NOTICE("You put [I] in [src]."))
+			if(LAZYLEN(oxygen_tanks) < 5)
 				update_icon()
-		if(href_list["hydrogen"])
-			if(hydrogentanks > 0)
-				var/obj/item/tank/hydrogen/P
-				if(hydtanks.len == hydrogentanks)
-					P = hydtanks[1]
-					hydtanks.Remove(P)
-				else
-					P = new /obj/item/tank/hydrogen(loc)
-				P.dropInto(loc)
-				to_chat(usr, "<span class='notice'>You take [P] out of [src].</span>")
-				hydrogentanks--
+		else
+			to_chat(user, SPAN_WARNING("\The [src] is full."))
+		updateUsrDialog()
+		return TRUE
+
+	if(istype(I, /obj/item/tank/hydrogen))
+		if(LAZYLEN(hydrogen_tanks) < 10)
+			if(!user.unEquip(I, src))
+				return
+			LAZYADD(hydrogen_tanks, I)
+			to_chat(user, SPAN_NOTICE("You put \the [I] in \the [src]."))
+			if(LAZYLEN(hydrogen_tanks) < 6)
 				update_icon()
-		add_fingerprint(usr)
+		else
+			to_chat(user, SPAN_WARNING("\The [src] is full."))
 		updateUsrDialog()
-	else
+		return TRUE
+
+	return ..()
+
+/obj/structure/tank_dispenser/Topic(href, href_list)
+
+	. = ..()
+	if(. || !usr || usr.incapacitated() || usr.restrained())
+		return
+
+	if(!Adjacent(usr))
 		close_browser(usr, "window=dispenser")
 		return
-	return
+
+	if(href_list["oxygen"] && LAZYLEN(oxygen_tanks))
+		var/obj/item/tank/O = oxygen_tanks[1]
+		LAZYREMOVE(oxygen_tanks, O)
+		O.dropInto(loc)
+		to_chat(usr, SPAN_NOTICE("You take \the [O] out of \the [src]."))
+		update_icon()
+		. = TOPIC_REFRESH
+
+	if(href_list["hydrogen"] && LAZYLEN(hydrogen_tanks))
+		var/obj/item/tank/H = hydrogen_tanks[1]
+		LAZYREMOVE(hydrogen_tanks, H)
+		H.dropInto(loc)
+		to_chat(usr, SPAN_NOTICE("You take \the [H] out of \the [src]."))
+		. = TOPIC_REFRESH
+
+	if(. == TOPIC_REFRESH)
+		usr.set_machine(src)
+		update_icon()
+		add_fingerprint(usr)
```

Comment:
Don't add fingerprints. Setting machine here is a giant hack but apparently I'm in a forgiving mood when it comes to the machine var at the moment.